service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -1021,7 +1021,9 @@ protected void onCommandConnectionEstablished(final HonoConnection commandConnec
             return Future.succeededFuture(registrationAssertion);
         }
 
-        final Future<String> gatewayId = getGatewayId(tenantId, deviceId, authenticatedDevice);
+        final Future<String> gatewayId = isGatewayMappingEnabled()
+                ? getGatewayId(tenantId, deviceId, authenticatedDevice)
+                : Future.succeededFuture(deviceId);
```

Comment:
my understanding is that this is an optimization in order to prevent the expensive invocation of the (remote) Device Connection service which would return the device ID anyway because that is what the HTTP adapter has mapped the (origin) device ID to in the first place. Is this correct?

Reply:
The entry in the Device Connection service needs to be set here because an incoming command message (on the `command/{tenant_id}` address) that is ultimately directed at a HTTP gateway could have first reached an MQTT adapter for example. The MQTT adapter would do the gateway mapping then.
That is why we can't just disable the gateway mapping logic in the HTTP adapter but instead have to make sure that the mapping logic does the job of mapping the command to the target device id, not the gateway in this case.